src/backend/distributed/transaction/worker_transaction.c

```diff
@@ -356,21 +358,38 @@ SendCommandListToWorkerOutsideTransaction(const char *nodeName, int32 nodePort,
  * any of the queries fail.
  */
 void
-SendCommandListToWorkerOutsideTransactionWithConnection(MultiConnection *workerConnection,
+SendCommandListToWorkerOutsideTransactionWithConnection(MultiConnection *connection,
 														List *commandList)
 {
-	MarkRemoteTransactionCritical(workerConnection);
-	RemoteTransactionBegin(workerConnection);
+	MarkRemoteTransactionCritical(connection);
+	RemoteTransactionBegin(connection);
+
+	List *concatenatedCommandList = list_copy(commandList);
+	bool isLoopBackConnection = false;
+	WorkerNode *workerNode = FindWorkerNode(connection->hostname, connection->port);
+
+	if (workerNode)
+	{
+		isLoopBackConnection = IsCoordinator() && workerNode->groupId ==
+							   GetLocalGroupId();
+	}
+
+	if (PropagateSessionSettingsForLoopbackConnection && isLoopBackConnection)
+	{
+		List *setCommandList = GetSetCommandListForNewConnections();
+		concatenatedCommandList = list_concat_copy(setCommandList,
+												   concatenatedCommandList);
```

Comment:
I find this implementation a little questionable. Perhaps it kind of sort of solves the immediate problem, but we're ignoring a lot of postgres idiosyncracies here. For instance, what happens if there are subtransactions that reset the settings on rollback? Do we reset it on the worker nodes too?

Perhaps we should properly implement SET propagation by extending the current SET LOCAL propagation. We can then decide whether we want to always propagate everything to localhost.

Reply:
Given that this is **outside** transaction function, and we either commit (or fail and rollback) should we care about sub-transactions?

Reply:
I find it quite hard to tell whether this utility function (SendCommandListToWorkerOutsideTransactionWithConnection) can be called from a subtransaction or not. It already has 9 different callers.

Reply:
ah, I guess you meant we will never need to propagate savepoint/rollback over this particular connection, so it's ok to use one static set of settings for the duration of the remote transaction. 

ok, it's a little hacky, but seems reasonable.

Reply:
> already has 9 different callers.

One another alternative we discussed with @gokhangulbiz was to kick this logic only in specific callers with a different function, namely during rebalancer and shard splits.

But @gokhangulbiz suggested that making it more generic is also an option, and wanted to open the PR to discuss that as well.

I kind of like the idea that "loop back - outside transaction" connections get the exact same settings as the current session running the command. That is probably what user wants/expects anyway.

We also restricted it to `IsCoordinator()` for now just to make the surface smaller, but also don't have a very specific reason for that restriction. I have not dived deep yet, but I expect only rebalancer and shard splits goes through this code-path?

Reply:
We decided not to merge this PR based on this discussion because this doesn't work for all the cases and doesn't fit well into the transaction management. We're going to keep #6396 for testing/debugging purposes and change this GUC setting's default value to ```off``` (See #6516) not to expose this feature publicly at this point.